addons/binding/org.openhab.binding.plugwise/src/main/java/org/openhab/binding/plugwise/internal/PlugwiseCommunicationHandler.java

```diff
@@ -0,0 +1,591 @@
+            true);
+    private final BlockingQueue<QueuedMessage> sentQueue = new ArrayBlockingQueue<>(MAX_BUFFER_SIZE, true);
+    private final BlockingQueue<Message> receivedQueue = new ArrayBlockingQueue<>(MAX_BUFFER_SIZE, true);
+
+    private final ReentrantLock sentQueueLock = new ReentrantLock();
+
+    // Background threads
+    private Thread messageProcessor;
+    private Thread messageSender;
+
+    // Serial communication fields
+    private ByteBuffer readBuffer = ByteBuffer.allocate(MAX_BUFFER_SIZE);
+    private SerialPort serialPort;
+    private WritableByteChannel outputChannel;
+    private int previousByte = -1;
+    private int sequentialWriteErrors = 0;
+
+    private boolean initialized = false;
+
+    public void addMessageListener(PlugwiseMessageListener listener) {
+        if (!isExistingMessageListener(listener)) {
+            filteredListeners.add(new FilteredMessageListener(listener));
+        }
+    }
+
+    public void addMessageListener(PlugwiseMessageListener listener, MACAddress macAddress) {
+        if (!isExistingMessageListener(listener, macAddress)) {
+            filteredListeners.add(new FilteredMessageListener(listener, macAddress));
+        }
+    }
+
+    /**
+     * Initialize this device and open the serial port
+     *
+     * @throws PlugwiseInitializationException if port can not be opened
+     */
+    @SuppressWarnings("rawtypes")
+    private void initializeSerialPort() throws PlugwiseInitializationException {
+
+        CommPortIdentifier portId = null;
+
+        // Parse ports and if the default port is found, initialized the reader
+        Enumeration portList = CommPortIdentifier.getPortIdentifiers();
+        while (portList.hasMoreElements()) {
+            CommPortIdentifier id = (CommPortIdentifier) portList.nextElement();
+            if (id.getPortType() == CommPortIdentifier.PORT_SERIAL
+                    && id.getName().equals(configuration.getSerialPort())) {
+                logger.debug("Serial port '{}' has been found", configuration.getSerialPort());
+                portId = id;
+                break;
+            }
+        }
+
+        if (portId == null) {
+            StringBuilder sb = new StringBuilder();
+            portList = CommPortIdentifier.getPortIdentifiers();
+            while (portList.hasMoreElements()) {
+                CommPortIdentifier id = (CommPortIdentifier) portList.nextElement();
+                if (id.getPortType() == CommPortIdentifier.PORT_SERIAL) {
+                    sb.append(String.format("%s%n", id.getName()));
+                }
+            }
+            throw new PlugwiseInitializationException(
+                    String.format("Serial port '%s' could not be found. Available ports are:%n%s",
+                            configuration.getSerialPort(), sb));
+        }
+
+        try {
+            serialPort = portId.open("openHAB", 2000);
```

Comment:
The binding does not use the label but I've made the label more unique anyhow, see: org.openhab.binding.plugwise.internal.PlugwiseCommunicationContext.initializeSerialPort()